src/python/pants/backend/python/typecheck/pyright/rules.py

```diff
@@ -194,6 +194,16 @@ async def pyright_typecheck_partition(
         ),
     )
 
+    # Force the requirements venv to materialize always by running a no-op.
```

Comment:
I understood the first sentence, but I don't think I fully understood this one:

> So this needs `cache_scope=ProcessCacheScope.PER_SESSION`.

I see that the definition says this:

```
# Will run once per Session, i.e. once per run of Pants. This happens because the engine
# de-duplicates identical work; the process is neither memoized in memory nor cached to disk.
PER_SESSION = "per_session"
```

Is the venv cached per-session as well? What am I missing?

Reply:
This is absolutely the key point, and it's important to clarify so we're all certain about the fix:

There are two caches in play here: The on-disk LMDB store which caches process executions and the on-disk venv cache, aka "named-caches". (There is also an in-memory cache of rule executions, but that is not impinging on the issue here.)

When running a pex process, the relevant underlying venv is either retrieved from cache, or created (and cached). However PyRight is special - it's a NodeJS binary that expects to operate against a Python virtualenv. So PyRight is *not* a pex process, and so doesn't create the venv, but assumes it was previously created. And without caches, it would be created on line 187, by the process that turns a PexRequest into a VenvPex. However, what if that process is cached in the LMDB store? It won't re-run. And so nothing ensures that the venv exists in the named-caches. If it does, we're fine, but if not, we blow up.

So the naive solution is to force that process to run every time, ignoring the cache, by setting cache_scope. But, that would potentially be expensive. So the more clever solution is the one you're implementing - run a separate no-op process that just ensures that the venv exists (and is cached), and then returns without doing anything, and so is fast. But we still have to force it to run every time! That is why it exists. 

Does that make sense?


And of course the test has to reflect this logic.

Reply:
Absolutely makes sense! And you unlocked some helpful knowledge for me on how to debug the rest of the failures in the test too. 

Thanks for the very clear and comprehensive explanation!